cosmwasm/contracts/wormchain-accounting/src/contract.rs

```diff
@@ -107,80 +107,93 @@ fn submit_observations(
+    let mut details = Vec::with_capacity(observations.len());
+    let mut events = Vec::with_capacity(observations.len());
+    for o in observations {
+        let key = transfer::Key::new(o.emitter_chain, o.emitter_address.into(), o.sequence);
+        let (status, event) =
             handle_observation(deps.branch(), o, guardian_set_index, quorum, signature)
-                .with_context(|| format!("failed to handle observation for key {key}"))
-        })
-        .filter_map(Result::transpose)
-        .collect::<anyhow::Result<Vec<_>>>()
-        .context("failed to handle `Observation`")?;
+                .with_context(|| format!("failed to handle observation for key {key}"))?;
+        details.push(TransferDetails {
+            key,
+            status: Some(status),
+        });
+        if let Some(evt) = event {
+            events.push(evt);
+        }
+    }
+
+    let data = to_binary(&details).context("failed to serialize transfer details")?;
 
     Ok(Response::new()
         .add_attribute("action", "submit_observations")
         .add_attribute("owner", info.sender)
+        .set_data(data)
         .add_events(events))
 }
 
 fn handle_observation(
     mut deps: DepsMut<WormholeQuery>,
     o: Observation,
     guardian_set_index: u32,
-    quorum: usize,
+    quorum: u32,
     sig: Signature,
-) -> anyhow::Result<Option<Event>> {
+) -> anyhow::Result<(TransferStatus, Option<Event>)> {
+    let digest = o.digest().context(ContractError::ObservationDigest)?;
+
     let digest_key = DIGESTS.key((o.emitter_chain, o.emitter_address.to_vec(), o.sequence));
+    let tx_key = transfer::Key::new(o.emitter_chain, o.emitter_address.into(), o.sequence);
+
     if let Some(saved_digest) = digest_key
         .may_load(deps.storage)
         .context("failed to load transfer digest")?
     {
-        let digest = o.digest().context(ContractError::ObservationDigest)?;
         if saved_digest != digest {
             bail!(ContractError::DigestMismatch);
         }
 
-        bail!(TransferError::DuplicateTransfer);
-    }
+        let data = accounting::query_transfer(deps.as_ref(), tx_key)
+            .context("failed to query transfer data")?;
 
-    let tx_key = transfer::Key::new(o.emitter_chain, o.emitter_address.into(), o.sequence);
+        return Ok((TransferStatus::Committed { data, digest }, None));
+    }
 
     let key = PENDING_TRANSFERS.key(tx_key.clone());
     let mut pending = key
         .may_load(deps.storage)
         .map(Option::unwrap_or_default)
         .context("failed to load `PendingTransfer`")?;
-    let data = match pending
-        .iter_mut()
-        .find(|d| d.guardian_set_index() == guardian_set_index && d.observation() == &o)
-    {
+    let data = match pending.iter_mut().find(|d| {
+        d.guardian_set_index() == guardian_set_index
+            && d.digest() == &digest
+            && d.tx_hash() == &o.tx_hash
```

Comment:
any chance you're not also matching for emitter chain here?

Reply:
The emitter chain is included as part of the digest calculation so I didn't feel like it was necessary to check it again (whereas the tx_hash is not part of the digest).